### PR TITLE
cli: add kata-overhead subcommand

### PR DESCRIPTION
**Pull request by jcvenegas**
### Is your feature request related to a problem? Please describe.

With the introduction of sandbox_cgroup_only configuration option for kata-runtime now is possible to calculate the overhead of a kata sandbox by calculating the differnce usage of the containers running in the guest and the resource of the Sandbox cgroup.

Consider add kata-runtime kata-overhead to provide overhead of kata for CPU and memory.

Having this will help users to calculate their own overhead uses-cases.

#### Describe the solution you'd like
```
kata-runtime kata-overhead  <sandbox-id|container-id>
cpu_overhead=120
memory_overhead=1000
```

### CODE REVIEW

**Comment by jcvenegas** [comment]
Related to: https://github.com/kata-containers/tests/pull/1975

File: cli/kata-overhead.go [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+		guestMemoryUsage += cs.CgroupStats.MemoryStats.Usage.Usage
+	}
+
+	hostMemoryUsage := finishSandboxStats.CgroupStats.MemoryStats.Usage.Usage
+	deltaTime := finishtTime - initTime
+
+	cpuUsageGuest := float64(guestFinalCPU-guestInitCPU) / float64(deltaTime) * 100
+	cpuUsageHost := float64(hostFinalCPU-hostInitCPU) / float64(deltaTime) * 100
+
+	fmt.Printf("cpu_guest=%f\n", cpuUsageGuest)
```
Comment (jodh-intel) [inline cli/kata-overhead.go]: Random thought: might be useful to offer a JSON output format option.

**Comment by bergwolf** [comment]
Could you add detailed documentation about the command output? Otherwise it LGTM. Thanks!

**Comment by chavafg** [comment]
@jcvenegas can you check the errors on this PR?
```
22:57:14 # github.com/kata-containers/runtime/cli [github.com/kata-containers/runtime/cli.test]
22:57:14 cli/main_test.go:83:6: cannot use testingImpl (type *vcmock.VCMock) as type virtcontainers.VC in assignment:
22:57:14 	*vcmock.VCMock does not implement virtcontainers.VC (wrong type for StatsSandbox method)
22:57:14 		have StatsSandbox("context".Context, string) (virtcontainers.ContainerStats, []virtcontainers.ContainerStats, error)
22:57:14 		want StatsSandbox("context".Context, string) (virtcontainers.SandboxStats, []virtcontainers.ContainerStats, error)
22:57:15 FAIL	github.com/kata-containers/runtime/cli [build failed]
```

File: virtcontainers/api.go [hunk 2]
```diff
@@ -758,6 +758,46 @@ func StatsContainer(ctx context.Context, sandboxID, containerID string) (Contain
 	return s.StatsContainer(containerID)
 }
 
+// StatsSandbox is the virtcontainers sandox stats entry point.
```
Comment (egernst) [inline virtcontainers/api.go]:
```suggestion
// StatsSandbox is the virtcontainers sandox stats entry point.
```
// StatsSandbox is the virtcontainers sandbox stats entry point.

File: cli/kata-overhead.go [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+	"github.com/urfave/cli"
+)
+
+var kataOverheadCLICommand = cli.Command{
+	Name:  "kata-overhead",
+	Usage: "provides kata overhead at sandbox level",
+	ArgsUsage: `<sandbox-id> [sandbox-id...]
+
+   <sandbox-id> is your name for the instance of the sandbox.`,
+	Description: `The overhead command shows the overehad of a running Kata Sandbox`,
```
Comment (egernst) [inline cli/kata-overhead.go]:
```suggestion
	Description: `The overhead command shows the overhead of a running Kata Sandbox`,
```

File: cli/kata-overhead.go [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+	Action: func(context *cli.Context) error {
+		ctx, err := cliContextToContext(context)
+		if err != nil {
+			return err
+		}
+
+		args := context.Args()
+		if !args.Present() {
+			return fmt.Errorf("Missing container ID, should at least provide one")
```
Comment (egernst) [inline cli/kata-overhead.go]: I think this shoudl be container ID?

```suggestion
			return fmt.Errorf("Missing sandbox ID, should at least provide one")
```

File: cli/kata-overhead.go [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+	span.SetTag("container", containerID)
+	span.SetTag("sandbox", sandboxID)
+
+	if status.State.State == types.StateStopped {
+		return fmt.Errorf("container with id %s is not running", status.ID)
+	}
+
+	initTime := time.Now().UnixNano()
+
+	initSandboxStats, initiContainersStats, err := vci.StatsSandbox(ctx, sandboxID)
```
Comment (egernst) [inline cli/kata-overhead.go]: 1) I think you mean initContainerStats?
2) initContainerStats reminds me of an init container.  Perhaps we can say initialSandboxStats and initialContainerStats instead? 

```suggestion
	initialSandboxStats, initialContainersStats, err := vci.StatsSandbox(ctx, sandboxID)
```

File: cli/kata-overhead.go [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+	if err != nil {
+		return err
+	}
+
+	hostInitCPU := initSandboxStats.CgroupStats.CPUStats.CPUUsage.TotalUsage
+	guestInitCPU := uint64(0)
```
Comment (egernst) [inline cli/kata-overhead.go]: 1) I think you mean initContainerStats?
2) initContainerStats reminds me of an init container.  Perhaps we can say initialSandboxStats and initialContainerStats instead?

File: cli/kata-overhead.go [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+	for _, cs := range initiContainersStats {
+		guestInitCPU += cs.CgroupStats.CPUStats.CPUUsage.TotalUsage
+	}
```
Comment (egernst) [inline cli/kata-overhead.go]: I think you mean initContainerStats?
initContainerStats reminds me of an init container. Perhaps we can say initialSandboxStats and initialContainerStats instead?

File: cli/kata-overhead.go (hunk 5 above)
Comment (egernst) [inline cli/kata-overhead.go]:
```suggestion
	initialSandboxStats, initialContainersStats, err := vci.StatsSandbox(ctx, sandboxID)
```

File: cli/kata-overhead.go [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+
+	initTime := time.Now().UnixNano()
+
+	initSandboxStats, initiContainersStats, err := vci.StatsSandbox(ctx, sandboxID)
+	if err != nil {
+		return err
+	}
+
+	hostInitCPU := initSandboxStats.CgroupStats.CPUStats.CPUUsage.TotalUsage
```
Comment (egernst) [inline cli/kata-overhead.go]:
```suggestion
	hostInitCPU := initialSandboxStats.CgroupStats.CPUStats.CPUUsage.TotalUsage
```

File: cli/kata-overhead.go [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+	guestInitCPU := uint64(0)
+	for _, cs := range initiContainersStats {
```
Comment (egernst) [inline cli/kata-overhead.go]:
```suggestion
	for _, cs := range initialContainersStats {
```

File: cli/kata-overhead.go [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+	fmt.Printf("cpu_guest_init=%d\n", guestInitCPU)
+	fmt.Printf("cpu_guest_final=%d\n", guestFinalCPU)
+	fmt.Printf("cpu_host=%f\n", cpuUsageHost)
+	fmt.Printf("cpu_host_init=%d\n", hostInitCPU)
+	fmt.Printf("cpu_host_final=%d\n", hostFinalCPU)
+	fmt.Printf("cpu_overhead=%f\n", cpuUsageHost-cpuUsageGuest)
+	fmt.Printf("memory_guest_bytes=%d\n", guestMemoryUsage)
+	fmt.Printf("memory_host_bytes=%d\n", hostMemoryUsage)
+	fmt.Printf("memory_overhead_bytes=%d\n", hostMemoryUsage-guestMemoryUsage)
+	fmt.Printf("vm_cpus=%d\n", finishSandboxStats.Cpus)
```
Comment (egernst) [inline cli/kata-overhead.go]: Is this too verbose? Perhaps we should pare it down for just what is useful for the user?

File: virtcontainers/sandbox.go [hunk 11]
```diff
@@ -1367,6 +1373,46 @@ func (s *Sandbox) StatsContainer(containerID string) (ContainerStats, error) {
...
+	if s.config.SandboxCgroupOnly {
+		// Override V1NoConstraints, if SandboxCgroupOnly is enabled
+		cgroupSubystems = cgroups.V1
+		path = s.state.CgroupPath
+		s.Logger().WithField("path", path).Debug("Deleting sandbox cgroups (all subsystems)")
+	} else {
+		path = cgroupNoConstraintsPath(s.state.CgroupPath)
+		s.Logger().WithField("path", path).Debug("Deleting no constraints cgroup")
+	}
+
```
Comment (egernst) [inline virtcontainers/sandbox.go]: cp/paste err?

File: virtcontainers/sandbox.go [hunk 12]
```diff
@@ -1367,6 +1373,46 @@ func (s *Sandbox) StatsContainer(containerID string) (ContainerStats, error) {
...
+	stats := SandboxStats{}
+
+	stats.CgroupStats.CPUStats.CPUUsage.TotalUsage = m.CPU.Usage.Total
+	stats.CgroupStats.MemoryStats.Usage.Usage = m.Memory.Usage.Usage
+	tids, err := s.hypervisor.getThreadIDs()
+	if err != nil {
+		return SandboxStats{}, err
+	}
+	stats.Cpus = len(tids.vcpus)
```
Comment (egernst) [inline virtcontainers/sandbox.go]: Is there a cheaper/easier way to get number of vCPUs?  Do we store this?

File: virtcontainers/sandbox.go [hunk 13]
```diff
@@ -62,6 +62,12 @@ type SandboxStatus struct {
 	Annotations map[string]string
 }
 
+// ContainerStats describes a container stats.
```
Comment (egernst) [inline virtcontainers/sandbox.go]:
```suggestion
// Sandbox describes a container stats.
```

File: virtcontainers/api.go [hunk 14]
```diff
@@ -758,6 +758,46 @@ func StatsContainer(ctx context.Context, sandboxID, containerID string) (Contain
...
+	}
+
+	containerStats := []ContainerStats{}
+	for _, c := range s.containers {
+		cstats, err := s.StatsContainer(c.id)
+		if err != nil {
+			return SandboxStats{}, []ContainerStats{}, err
+		}
+		containerStats = append(containerStats, cstats)
+	}
```
Comment (egernst) [inline virtcontainers/api.go]: Is there much utility in returning each container stat set individually?
If a user wanted this, they could use the existing API?
If this is desired, should we consider returning the some of container stats?  WDYT?

Comment (egernst) [reply virtcontainers/api.go]: I'm okay with it as is. We can always sum on the usage side (CLI), and that way we don't lose any potentially interesting information.

**Comment by egernst** [comment]
repushed with squashed/cleanup commits (no changes)

**Comment by egernst** [comment]
Both debian and fedora seem to fail with 
```
19:49:08 Failed at 171: status=$(get_jaeger_status "$service")
19:49:11 jaeger
19:49:11 jaeger
``` 
Which does not look relevant to this PR at all.

**Pull request merged**